(WiP): [Not-yet-]stable branch to parcel out for upstream and VV

brpocock opened the pull request:
There are a lot of feature branches that are merged down here.

As I post this PR, there are still serious errors that prevent it from working, but the goal is to make this a good landable branch with the vast majority of the improvements, such as the top-level build work, and the sketchy outline of a type system. The goal, of course, is to be able to send it downstream (for Tootsville) and start parceling out chunks for upstream as well.

In other words, all of the works in progress are being landed onto this huge mess of a branch and then the outstanding heap of issues can be reconsidered in light of what remains.

davazp commented:
wow! There are some very interesting things in there!

Is there any way we can make easier for you to apply your changes incrementally?

Most of the time I will apply the PR quickly even if it not really completed. 

Or maybe I don't understand well how it works., Is your fit workflow documented somewhere?

vendethiel commented on src/compiler/compile-file.lisp:
```diff
@@ -78,6 +78,12 @@ execute BODY. Also binds LAST-FORM and FORM-COUNT."
+ *  ⸨☕λ⸩ Compiled by 𝓙𝓢ℂ𝕃
+ * ~@[(Romance Ⅱ fork) ~]version ~a, Git commit ~a
+ * Source file: ~a */" 
+              *version* (violet-volts-p)
```
`(violet-volts-p)`?

brpocock commented:
Apologies in advance for a meandering response:

Right now I've horribly broken the world 😥 but I'm struggling to get the
types.lisp code to work. It's terribly circular with defstruct.lisp; if you
try (load "jscl") (bootstrap-core) you'll see it burn down, I think
probably first with something like type-error "#s(foo) is not type foo"
(probably in lambda-list.lisp, I think) — I have run against several dead
ends trying to resolve that. I hope there are only a few more dead-ends
before I find the solution.

The pushes I've been sending up under "WiP" are mostly daily backups of
"what i did while watching TV" lately and not exactly tidy commits at all …
I figure I'd squash the branch and do a "clean" diff against master … I'm
afraid that's all the documentation of my workflow right now!

My first/next goal is to get it to build and pass at least the existing
tests, and add some new ones. Then merge this Work-in-progress branch down
and start some feature branches off your master, more or less by cut and
paste from mine, ensure they work in those branches and PR them back to
you.

I externed-in the GCL tests with some longer term thought of using them as
a nice independent gauge of ANSI-ness. I've done nothing really to
integrate them, though.

There are a smattering of changes to try to support compilers other than
SBCL. Quasiquotes/Backquotes from READ are particularly implementation
dependent. With the understanding that almost none of them really work,
those that exist could be committed.

I suspect some of the compiler work is safe to parcel out without the type
system changes but i haven't tried.

A point of concern I have is that the "FSET" style bindings don't honour
the *environment*/BINDING and I suspect it maybe should be
removed/consolidated? … in particular I added a hack that (fdefinition
(setf foo)) is bound on symbol foo in a setf slot, but then also tried to
mimic FSET as FSET-SETF for globals, and that seems wrong. Perhaps because
the NIL environment wasn't being shared (on my branch as
*global-environment*) that was standing in for it? — those bits landed in
the compiler code as well as boot.lisp I think. (&Whole and & Environment
may work now, and macro-functions should take the two-arg form. The whole
world of "kinds of lambda lists" in various contexts is probably often
incorrect right now, though, as the several different flavours of lambda
lists are all run together.)

The "compile-file" bits moved out from jscl.lisp and factored out a bit to
make building external app packages a little easier and also could probably
be split off into a feature branch. There's a bit of æsthetics and a bit of
real functionality rather interspersed.

Sprinkled throughout, I moved many comments into docstrings and added a
documentation target (requiring DECLT) to try to make a nice PDF. The ASDF
file exists for now only to convince DECLT what to do.

The print.lisp updates are probably also candidates to parcel out. They do
build upon the Unicode branch , though, which i think is still floating as
a PR? I know it's rather hefty but I think that PR/branch is fairly solid,
if incomplete. The new format code should have letter-case ~( and list ~{
working, as well as ~@r Roman numerals, and ~/funcall/. I had started on ~[
but don't think I've committed that. ~< justification I may throw in, but
the advanced ~< forms confuse me enough I may not get to them for a long
while.

I had also "repackaged" a few things into JSCL/FFI and less into JSCL/CLTL2
and JSCL/MOP (one or two functions each) that were needed for the type
system and (interestingly) LOOP.

As best I can think of at the moment those are the main things. The "type
system" work particularly is rather invasive and if you grep
(TODO|HACK|FIXME) there are many. If I can get it stable enough to build
and at least run some tests the cleaning up of residual bugs will probably
be quite shallow I hope.


Longer term, I don't think I've landed much of it, but I have started on a
very minimal DefGeneric and a rather stupid DefClass based on the same code
as DefStruct mostly; I don't anticipate a full CLOS, but enough that some
common things would work. You may see that I changed the storage vector
type to hold type/class metadata as its "kind" in anticipation of that, and
also for array dimensions.

Particularly Print-Object, Documentation, and a subset of Gray Streams for
(web console, Node console, XHR, and WebRTC data streams) seem to require
generic functions.


On Sun, Dec 4, 2016, 01:43 David Vázquez Púa <notifications@github.com>
wrote:

wow! There are some very interesting things in there!

Is there any way we can make easier for you to apply your changes
incrementally?

Most of the time I will apply the PR quickly even if it not really
completed.

Or maybe I don't understand well how it works., Is your fit workflow
documented somewhere?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/jscl-project/jscl/pull/263#issuecomment-264687717>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AB-HO1IFvuSlEjg8SjTxI34Q-63fzsjyks5rEmEHgaJpZM4K6JlF>
.

brpocock commented:
That's meant to identify the romance-ii/JSCL fork distinct from
jscl-project/JSCL — a "tainted" version compared to the canonical one.

It's presently just (defun violet-volts-p () T) but I mean to do something
more clever…


On Tue, Dec 6, 2016, 04:15 ven <notifications@github.com> wrote:

*@vendethiel* commented on this pull request.
------------------------------

In src/compiler/compile-file.lisp
<https://github.com/jscl-project/jscl/pull/263#pullrequestreview-11546765>:

> @@ -78,6 +78,12 @@ execute BODY. Also binds LAST-FORM and FORM-COUNT."
 forms if PRINT is set."
   (with-compilation-restarts (filename)
     (with-compile-file-bindings (filename :verbosep print)
+      (format out "~&/** @preserve
+ *  ⸨☕λ⸩ Compiled by 𝓙𝓢ℂ𝕃
+ * ~@[(Romance Ⅱ fork) ~]version ~a, Git commit ~a
+ * Source file: ~a */"
+              *version* (violet-volts-p)

(violet-volts-p)?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/jscl-project/jscl/pull/263#pullrequestreview-11546765>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AB-HO6qNxNefRcZVVjWjkgbPIkrg75Lgks5rFSeggaJpZM4K6JlF>
.

vendethiel commented:
> That's meant to identify the romance-ii/JSCL fork distinct from
> jscl-project/JSCL — a "tainted" version compared to the canonical one.
> 
> It's presently just (defun violet-volts-p () T) but I mean to do something
> more clever…

I'm sorry if I'm being really dense, but shouldn't this:

```
+      (format out "~&/** @preserve
+ *  ⸨☕λ⸩ Compiled by 𝓙𝓢ℂ𝕃
+ * ~@[(Romance Ⅱ fork) ~]version ~a, Git commit ~a
+ * Source file: ~a */"
+              *version* (violet-volts-p)
```

be this instead? (order swapped)

```
+      (format out "~&/** @preserve
+ *  ⸨☕λ⸩ Compiled by 𝓙𝓢ℂ𝕃
+ * ~@[(Romance Ⅱ fork) ~]version ~a, Git commit ~a
+ * Source file: ~a */"
+              (violet-volts-p) *version*
```

brpocock commented:
Totally correct, my mistake! Thanks.

On Tue, Dec 6, 2016, 10:55 ven <notifications@github.com> wrote:

> That's meant to identify the romance-ii/JSCL fork distinct from
> jscl-project/JSCL — a "tainted" version compared to the canonical one.
>
> It's presently just (defun violet-volts-p () T) but I mean to do something
> more clever…
>
> I'm sorry if I'm being really dense, but shouldn't this:
>
> +      (format out "~&/** @preserve
> + *  ⸨☕λ⸩ Compiled by 𝓙𝓢ℂ𝕃
> + * ~@[(Romance Ⅱ fork) ~]version ~a, Git commit ~a
> + * Source file: ~a */"
> +              *version* (violet-volts-p)
>
> be this instead? (order swapped)
>
> +      (format out "~&/** @preserve
> + *  ⸨☕λ⸩ Compiled by 𝓙𝓢ℂ𝕃
> + * ~@[(Romance Ⅱ fork) ~]version ~a, Git commit ~a
> + * Source file: ~a */"
> +              (violet-volts-p) *version*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jscl-project/jscl/pull/263#issuecomment-265187338>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB-HO2oCfTDZZ7TNlGI3Y3KgkGoPwzycks5rFYWGgaJpZM4K6JlF>
> .
>

brpocock commented:
I have flattened the commit history a great deal and tried to make the commit message give an overall review of the latest changes that are on this branch. This should make it easier for me to parcel them out. into patch-branches and real pull requests. It won't happen overnight yet, and things are still quite broken.

brpocock commented:
In case anyone is interested in this branch: No, I didn't get it back working in 12/2016 as the branch name seems to promise ☹ but it's “logarithmicly” approaching a stable check-point.

The major “late” change was to split up the JSCL package into (rather numerous) packages, including **JSCL/Bootstrap** to contain the load and bootstrap functions, and **JSCL/Common-Lisp** to contain the cross-compilation.

The bootstrap procedure is therefor now:

- Compile JSCL and its CL package to native code, including the functions that would be provided by **prelude.js** are instead emulated with native code.
- Using that “hosted” version of JSCL, then re-load (mostly) the same files, but instead, write out jscl.js by using the **JSCL/CL:Compile-File** function (running in “hosted” mode), which creates JavaScript rather than native binary code.

The package-renaming and JavaScript-low-level (**prelude**) emulation parts are still slightly broken. I'm hunting down these bugs one-by-one. I expect (pessimistically) there to be other errors which will then be exposed once the compilation can complete successfully, at which point, I'll hunt them down with the assistance of the test suites, and revisit some remaining TODO and FIXME comments as well. Then, of course, I can begin trying to “chunk” more realistically-sized pull requests out from this mass for upstream review/integration.

To gloss over, the major items are (in random order)
- The UTF-8 fixes and so forth from PR #247 
- Major redefinition of packages to modularize things
- Several small fixes to ANSI CL compliance, including new Format handlers
- DefStruct using “typed” vectors, including merged (and altered) version of #260 
- a minimally correct CL “Type System” with Type-Of, SubtypeP, TypeCase, etc.
- The barest minimal outlines of DefClass and DefGeneric beginning to be sketched-in
- A temporary non-CLOS implementation of Streams in a minimally-generic way, meant to move up to Gray Streams once DefGeneric works better
- Some sketches toward CLOS-based Conditions, but nothing very useful yet.
- Breaking up Compile-File and Bootstrap to be potentially more useful for applications other than the REPL demos
- Documentation “improvements” (subjectively) including migrating many comments to docstrings
- Some tweaks to the “Build System” itself that work better for me, but perhaps not others; including a Makefile and renaming the Node.js REPL executable.
- Some rather low-level changes to handle SetF functions and FDefinition (As in ANSI, but not in CLtL1)
- Some improvements on Arrays around dimensions and fill-pointers, particularly
- A few changes to **prelude.js** for compatibility with Google Closure Compiler minification, including **package** keyword conflicts and ensuring that “exported” names are not minified
- Some improvements on JSCL/FFI (JavaScript FFI) and a simplified (experimental) notation for “unquoted” variables in the ♯j reader, in contrast to (but similar intent as) #262 
- A great deal of changes (and additions) to the Compat functions to allow running JSCL itself more-or-less fully in “hosted” environment, using Hash-Table to represent JS Object maps.

See the rather much longer e-mail comment message (on https://github.com/jscl-project/jscl/pull/263) from early December for perhaps a better enumeration of the features merged down here.

davazp commented:
Can we close this PR until we are ready to rebase and (if possible) propose smaller ones? 

I don't like having an pull request open forever without the possibility of doing any action of it.

Maybe you can open an issue to plan how to integrate this. If it is not practical to split in smaller commits we can push it into a branch and start integrating both branches slowly. Anyway, please open an issue for it and close this PR for now.

davazp commented:
@brpocock I'm closing this PR as it has been open for a long time now.  Please don't hesistate to re-open it if there is updates!
 
I would be very happy to integrate many of your changes into master but at this point it is not really feasible.
